src/storage/recovery/recovery_manager.cpp

```diff
@@ -145,6 +144,14 @@ void RecoveryManager::ProcessCommittedTransaction(noisepage::transaction::timest
 
   // Commit the txn
   txn_manager_->Commit(txn, transaction::TransactionUtil::EmptyCallback, nullptr);
+
+  last_applied_txn_id_ = std::max(last_applied_txn_id_, txn_id);
+  if (replication_manager_ != DISABLED) {
+    // Replicas have to send back their list of deferred transactions that were processed, periodically.
+    if (replication_manager_->IsReplica()) {
+      replication_manager_->GetAsReplica()->NotifyPrimaryTransactionApplied(txn_id);
```

Comment:
Not for this PR, but in the future, it may be worth looking into sending back transaction ids to the primary in batches. Since we're usually applying multiple deferred transactions at once, it might be more efficient.

Reply:
Sounds good, added a todo. Might be particularly tricky to find a good batch size with sync replication / if we care about latency instead of throughput.